lib/rewrite/rewrite-set-severity.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2019 Balabit
+ * Copyright (c) 2019 Kokan <kokaipeter@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "rewrite-set-severity.h"
+#include "template/templates.h"
+#include "syslog-names.h"
+#include "scratch-buffers.h"
+
+#include <stdlib.h>
+
+typedef struct _LogRewriteSetSeverity LogRewriteSetSeverity;
+
+struct _LogRewriteSetSeverity
+{
+  LogRewrite super;
+  LogTemplate *severity;
+};
+
+static gint
+_convert_severity_as_number(GString *severity_text)
+{
+  char *endptr;
+  long int severity = strtol(severity_text->str, &endptr, 10);
+  if (0 == endptr)
+    return -1;
+
+  if (endptr[0] != '\0')
+    return -1;
+
+  if (severity > 7)
+    return -1;
+
+  return severity;
+}
+
+static gint
+_convert_severity_as_text(GString *severity_text)
+{
+  return syslog_name_lookup_level_by_name(severity_text->str);
+}
+
+static gint
+_convert_severity(GString *severity_text)
+{
+  gint severity = _convert_severity_as_number(severity_text);
```

Comment:
Templates evaluated to an empty string will set the highest `emergency` severity by mistake.
manpage:
> If there were no digits at all,  strtol()  stores the original value of nptr in *endptr (and returns 0).

You have a check for `(endptr[0] != '\0')`, but unfortunately, the original value is also `\0` in this case.

I've added the following test case to check this behavior:
<details>
  <summary>Click to expand!</summary>

```
Test(set_severity, test_set_severity_with_template_evaluating_to_empty_string)
{
  debug_flag = 1;

  LogPathOptions path_options = LOG_PATH_OPTIONS_INIT;
  LogMessage *msg = log_msg_new_empty();

  int default_severity = msg->pri & LOG_PRIMASK;

  LogRewrite *set_severity = log_rewrite_set_severity_new(_create_template("${nonexistentvalue}"), cfg);

  log_pipe_init(&set_severity->super);
  log_msg_ref(msg);
  log_pipe_queue(&set_severity->super, msg, &path_options);

  cr_assert_eq(msg->pri & LOG_PRIMASK, default_severity, "empty templates should not change the original severity");
  assert_grabbed_log_contains("invalid severity to set");

  log_msg_unref(msg);
  log_pipe_deinit(&set_severity->super);
  log_pipe_unref(&set_severity->super);
}
```
</details>

Reply:
Good catch!  Let's compare the addresses then…

Reply:
Thanks!